pkg/gardenlet/controller/federatedseed/seed_control.go

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package federatedseed
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	gardencoreinformers "github.com/gardener/gardener/pkg/client/core/informers/externalversions"
+	gardencorelisters "github.com/gardener/gardener/pkg/client/core/listers/core/v1alpha1"
+	extensionsclientset "github.com/gardener/gardener/pkg/client/extensions/clientset/versioned"
+	extensionsinformers "github.com/gardener/gardener/pkg/client/extensions/informers/externalversions"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/controllerutils"
+
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	confighelper "github.com/gardener/gardener/pkg/gardenlet/apis/config/helper"
+	"github.com/gardener/gardener/pkg/gardenlet/controller/federatedseed/shootstate"
+	"github.com/gardener/gardener/pkg/logger"
+	seedpkg "github.com/gardener/gardener/pkg/operation/seed"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// Controller is responsible for maintaining multiple federated Seeds' controllers
+type Controller struct {
+	federatedSeedControllerManager *federatedSeedControllerManager
+	k8sGardenClient                kubernetes.Interface
+
+	config *config.GardenletConfiguration
+
+	seedLister    gardencorelisters.SeedLister
+	seedHasSynced cache.InformerSynced
+	seedQueue     workqueue.RateLimitingInterface
+
+	recorder record.EventRecorder
+
+	ctx                    context.Context
+	numberOfRunningWorkers int
+	workerCh               chan int
+}
+
+// NewFederatedSeedController creates new controller that syncs extensions states to the ShootState resource in the garden cluster.
+func NewFederatedSeedController(ctx context.Context, k8sGardenClient kubernetes.Interface, gardenCoreInformerFactory gardencoreinformers.SharedInformerFactory, config *config.GardenletConfiguration, recorder record.EventRecorder) *Controller {
```

Comment:
Just a note: if you want to make this FederatedSeedController more generic we might have to add functionality to support custom callbacks for SeedAdd SeedUpdate and SeedDelete because different conditions for acting on seed events could be required by different federated controllers. Same goes for the reconciliation function used to process seeds in the worker's `seedQueue`.

Additionally, the `federatedSeedControllerManager` has to be made more generic (currently it only supports SyncControllers) 

There are other considerations to take into account but we can discuss this into more detail internally.